File: backend-go/internal/httpapi/auth_handlers.go
package httpapi

import (
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/danielraab/na-notes-app/backend-go/internal/apperr"
)

// isSafeRedirectPath restricts post-login redirects to an in-app path, to
// avoid the login flow being used as an open redirect.
func isSafeRedirectPath(p string) bool {
	return p != "" && strings.HasPrefix(p, "/") && !strings.HasPrefix(p, "//")
}

func (d *Deps) handleLogin(w http.ResponseWriter, r *http.Request) {
	redirectTo := r.URL.Query().Get("redirectTo")
	if !isSafeRedirectPath(redirectTo) {
		redirectTo = "/"
	}

	req, err := d.AuthStore.CreateOIDCRequest(r.Context(), redirectTo)
	if err != nil {
		respondDomainError(w, err)
		return
	}

	http.Redirect(w, r, d.OIDC.AuthCodeURL(req.State, req.CodeVerifier), http.StatusFound)
}

func (d *Deps) handleCallback(w http.ResponseWriter, r *http.Request) {
	code := r.URL.Query().Get("code")
	state := r.URL.Query().Get("state")
	if code == "" || state == "" {
		respondError(w, http.StatusBadRequest, "VALIDATION_ERROR", "missing code or state")
		return
	}

	req, err := d.AuthStore.ConsumeOIDCRequest(r.Context(), state)
	if err != nil {
		respondError(w, http.StatusBadRequest, "INVALID_STATE", "login request expired or was already used")
		return
	}

	claims, err := d.OIDC.Exchange(r.Context(), code, req.CodeVerifier)
	if err != nil {
		slog.Error("oidc exchange failed", "error", err)
		respondError(w, http.StatusBadGateway, "OIDC_EXCHANGE_FAILED", "could not complete login with identity provider")
		return
	}

	user, err := d.Users.UpsertFromOIDC(r.Context(), claims.Subject, claims.Email, claims.DisplayName, claims.AvatarURL)
	if err != nil {
		respondDomainError(w, err)
		return
	}

	session, err := d.AuthStore.CreateSession(r.Context(), user.ID)
	if err != nil {
		respondDomainError(w, err)
		return
	}

	secure := strings.HasPrefix(d.Config.PublicBaseURL, "https://")
	maxAge := int(time.Until(session.ExpiresAt).Seconds())

	http.SetCookie(w, &http.Cookie{
		Name:     SessionCookieName,
		Value:    session.ID,
		Path:     "/",
		Domain:   d.Config.CookieDomain,
		HttpOnly: true,
		Secure:   secure,
		SameSite: http.SameSiteLaxMode,
		MaxAge:   maxAge,
	})
	// Readable by frontend JS on purpose — it's echoed back as the
	// X-CSRF-Token header, never trusted as an identity credential itself.
	// Domain defaults to "" (host-only cookie); set COOKIE_DOMAIN when the
	// frontend and backend are on different subdomains of the same parent
	// domain, or the frontend's JS can never read this cookie to echo it.
	http.SetCookie(w, &http.Cookie{
		Name:     CSRFCookieName,
		Value:    session.CSRFToken,
		Path:     "/",
		Domain:   d.Config.CookieDomain,
		HttpOnly: false,
		Secure:   secure,
		SameSite: http.SameSiteLaxMode,
		MaxAge:   maxAge,
	})

	http.Redirect(w, r, d.Config.FrontendURL+req.RedirectTo, http.StatusFound)
}

func (d *Deps) handleLogout(w http.ResponseWriter, r *http.Request, userID string) {
	if cookie, err := r.Cookie(SessionCookieName); err == nil {
		_ = d.AuthStore.DeleteSession(r.Context(), cookie.Value)
	}
	clearCookie(w, SessionCookieName, d.Config.CookieDomain)
	clearCookie(w, CSRFCookieName, d.Config.CookieDomain)
	respondNoContent(w)
}

func clearCookie(w http.ResponseWriter, name, domain string) {
	http.SetCookie(w, &http.Cookie{Name: name, Value: "", Path: "/", Domain: domain, MaxAge: -1})
}

func (d *Deps) handleMe(w http.ResponseWriter, r *http.Request, userID string) {
	if userID == "" {
		respondError(w, http.StatusUnauthorized, "UNAUTHENTICATED", "login required")
		return
	}
	user, err := d.Users.GetByID(r.Context(), userID)
	if err != nil {
		respondDomainError(w, apperr.ErrNotFound)
		return
	}
	respondJSON(w, http.StatusOK, toUserDTO(user))
}

File: backend-rust/src/httpapi/auth_handlers.rs
use std::sync::Arc;

use axum::body::Body;
use axum::extract::{Query, State};
use axum::http::{header, HeaderMap, StatusCode};
use axum::response::Response;
use axum::Json;
use serde::Deserialize;

use crate::apperr::AppError;

use super::cookies::{
    clear_cookie, read_cookie, set_cookie, CookieAttrs, CSRF_COOKIE_NAME, SESSION_COOKIE_NAME,
};
use super::dto::to_user_dto;
use super::middleware::{OptionalAuth, RequireAuth};
use super::respond::ApiError;
use super::server::Deps;

/// Restricts post-login redirects to an in-app path, to avoid the login
/// flow being used as an open redirect.
fn is_safe_redirect_path(p: &str) -> bool {
    !p.is_empty() && p.starts_with('/') && !p.starts_with("//")
}

#[derive(Deserialize)]
pub struct LoginQuery {
    #[serde(default, rename = "redirectTo")]
    redirect_to: String,
}

pub async fn handle_login(
    State(deps): State<Arc<Deps>>,
    Query(q): Query<LoginQuery>,
) -> Result<Response, ApiError> {
    let redirect_to = if is_safe_redirect_path(&q.redirect_to) {
        q.redirect_to
    } else {
        "/".to_string()
    };

    let req = deps.auth_store.create_oidc_request(redirect_to).await?;
    let url = deps.oidc.auth_code_url(req.state, &req.code_verifier);

    Ok(redirect_found(&url))
}

#[derive(Deserialize)]
pub struct CallbackQuery {
    #[serde(default)]
    code: String,
    #[serde(default)]
    state: String,
}

pub async fn handle_callback(
    State(deps): State<Arc<Deps>>,
    Query(q): Query<CallbackQuery>,
) -> Result<Response, ApiError> {
    if q.code.is_empty() || q.state.is_empty() {
        return Err(ApiError::new(
            StatusCode::BAD_REQUEST,
            "VALIDATION_ERROR",
            "missing code or state",
        ));
    }

    let req = deps
        .auth_store
        .consume_oidc_request(q.state)
        .await
        .map_err(|_| {
            ApiError::new(
                StatusCode::BAD_REQUEST,
                "INVALID_STATE",
                "login request expired or was already used",
            )
        })?;

    let claims = deps
        .oidc
        .exchange(q.code, req.code_verifier)
        .await
        .map_err(|e| {
            tracing::error!(error = %e, "oidc exchange failed");
            ApiError::new(
                StatusCode::BAD_GATEWAY,
                "OIDC_EXCHANGE_FAILED",
                "could not complete login with identity provider",
            )
        })?;

    let user = deps
        .users
        .upsert_from_oidc(
            claims.subject,
            claims.email,
            claims.display_name,
            claims.avatar_url,
        )
        .await?;
    let session = deps.auth_store.create_session(user.id).await?;

    let secure = deps.config.public_base_url.starts_with("https://");
    let max_age = (session.expires_at - chrono::Utc::now()).num_seconds();
    let domain = deps.config.cookie_domain.as_deref();

    let session_cookie = set_cookie(
        SESSION_COOKIE_NAME,
        &session.id,
        domain,
        CookieAttrs {
            http_only: true,
            secure,
            max_age: Some(max_age),
        },
    );
    // Readable by frontend JS on purpose — it's echoed back as the
    // X-CSRF-Token header, never trusted as an identity credential itself.
    let csrf_cookie = set_cookie(
        CSRF_COOKIE_NAME,
        &session.csrf_token,
        domain,
        CookieAttrs {
            http_only: false,
            secure,
            max_age: Some(max_age),
        },
    );

    let location = format!("{}{}", deps.config.frontend_url, req.redirect_to);
    let response = Response::builder()
        .status(StatusCode::FOUND)
        .header(header::LOCATION, location)
        .header(header::SET_COOKIE, session_cookie)
        .header(header::SET_COOKIE, csrf_cookie)
        .body(Body::empty())
        .expect("valid response");

    Ok(response)
}

pub async fn handle_logout(
    State(deps): State<Arc<Deps>>,
    _auth: RequireAuth,
    headers: HeaderMap,
) -> Result<Response, ApiError> {
    if let Some(session_id) = read_cookie(&headers, SESSION_COOKIE_NAME) {
        let _ = deps.auth_store.delete_session(session_id).await;
    }
    let domain = deps.config.cookie_domain.as_deref();
    let response = Response::builder()
        .status(StatusCode::NO_CONTENT)
        .header(
            header::SET_COOKIE,
            clear_cookie(SESSION_COOKIE_NAME, domain),
        )
        .header(header::SET_COOKIE, clear_cookie(CSRF_COOKIE_NAME, domain))
        .body(Body::empty())
        .expect("valid response");
    Ok(response)
}

pub async fn handle_me(
    State(deps): State<Arc<Deps>>,
    OptionalAuth(user_id): OptionalAuth,
) -> Result<Json<super::dto::UserDto>, ApiError> {
    let Some(user_id) = user_id else {
        return Err(ApiError::unauthenticated());
    };
    let user = deps
        .users
        .get_by_id(user_id)
        .await
        .map_err(|_| AppError::NotFound)?;
    Ok(Json(to_user_dto(user)))
}

fn redirect_found(url: &str) -> Response {
    Response::builder()
        .status(StatusCode::FOUND)
        .header(header::LOCATION, url)
        .body(Body::empty())
        .expect("valid response")
}
